File: solutions/20-valid_parentheses/cpp/solution.cpp
/*
 * Problem 2. Valid Parenthesis
 * Author: Artur Gasparyan
 * Date: 30 Dec 2022
 * Link: https://leetcode.com/problems/valid-parentheses/submissions/868249682/
*/

class Solution {
public:
    bool isValid(string s) {
        std::stack<char> closerStack;

        for (int i = 0; i < s.length(); ++i) {
            char c = s[i];

            if (isOpener(c)) {
                closerStack.push(c);

            }
            else {
                if (closerStack.empty() || closerStack.top() != getOpener(c))
                    return false;
                    
                closerStack.pop();
            }
        }

        return closerStack.empty();
    }


private:
    const bool isOpener(const char c) {
        return c == '(' || c == '[' || c == '{'; 
    }

    const char getOpener(const char c) {
        switch (c) {
            case ')':
                return '(';
            case ']':
                return '[';
            case '}':
                return '{';
            default:
                return 'X';
        }
    }
};

File: solutions/20-valid_parentheses/python/solution.py
""" Problem 20. Valid Parenthesis
    Author: Artur Gasparyan
    Date: 21 Jan 2023
    Link: https://leetcode.com/problems/valid-parentheses/submissions/1025677980/
"""


class Solution:
    @staticmethod
    def isValid(s):
        stack = []
        for c in s:
            if c == "(" or c == "{" or c == "[":
                stack.append(c)
            else:
                if not stack:
                    return False

                m = stack.pop(-1)
                if (
                    (m == "(" and c != ")")
                    or (m == "[" and c != "]")
                    or (m == "{" and c != "}")
                ):
                    return False

        return not stack
